clang/lib/CodeGen/CodeGenFunction.cpp

```diff
@@ -748,8 +748,16 @@ void CodeGenFunction::StartFunction(GlobalDecl GD, QualType RetTy,
     const bool SanitizeBounds = SanOpts.hasOneOf(SanitizerKind::Bounds);
     SanitizerMask no_sanitize_mask;
     bool NoSanitizeCoverage = false;
+    bool NoSanitizeRealtime = false;
 
     for (auto *Attr : D->specific_attrs<NoSanitizeAttr>()) {
+      // The realtime sanitizer is not handled by SanOpts
+      // if it is enabled, it should remain enabled
+      if (Attr->hasRealtime()) {
+        NoSanitizeRealtime = true;
+        continue;
+      }
```

Comment:
Alright, this feels like a kind of hack here.

Basically the idea is what I described in the original description of this PR. On line 1573, if the realtime sanitizer is disabled, it is turned off. That sounds stupid to type out, but it's important.

If the realtime sanitizer is turned off, the block on line 1573 is never tripped, and we never get the opportunity to add the bypass call in.


Our options:

1. Don't respect the __nosanitize attribute, which would mean we just remove this entire change in this function
2. do this some other way?? Not sure how...


The psuedocode (implemented here) feels hacky:


This feels a little hacky because it's essentially saying "Do not ACTUALLY turn off the RADSan if the flag tells you to". This means that while SanOpts will be accurate for all other sanitizers, it will not be accurate for RADSan.

Reply:
After having a think, this doesn't feel that bad to me. Fundamentally what "no sanitize" means for a RADSan-sanitized function is different to what "no sanitize" means in other sanitizers. For RADSan, it's not really "turning it off" - it just means that we have to do something different: inserting the `radsan_off()` and `radsan_on()` calls! I'm not sure there's another way, as you imply as well. With a clear enough comment, I feel like this is a reasonable approach!